src/scaffoldmaker/meshtypes/meshtype_3d_bone1.py

```diff
@@ -0,0 +1,461 @@
+#                        n1c = 1
+#                    elif n1 == n2:
+#                        n1c = elementsCountAcross[1] - 1
+#                if cylinder1._shield.nodeId[0][n2c][n1c]:
+#                    nodesIdCylinderProximalEnd.append(cylinder1._shield.nodeId[0][n2c][n1c])
+
+        # ******************************************************************************************************************************
+        # generate hemisphere extra nodes.
+        rangeOfRequiredElements = [[0, elementsCountAcross[0]], [0, elementsCountAcross[1]],
+                                   [0, int(0.5 * elementsCountAcross[2]) - 1]]
+        radius_scale = options['Lower scale']
+        for n2 in range(elementsCountAcross[0] + 1):
+            for n1 in range(elementsCountAcross[1] + 1):
+                if hemisphere.px[0][n2][n1]:
+                    x = hemisphere.px[0][n2][n1]
+                    hemisphere.px[0][n2][n1] = [radius_scale * x[0], radius_scale * x[1], x[2]]
+        #
+        nodeIdentifier = max(1, getMaximumNodeIdentifier(nodes) + 1)
+        nodeIdentifier = hemisphere.generateNodes(fm, coordinates, nodeIdentifier,
+                                                  rangeOfRequiredElements)
+
+        # replace the hemispherical nodes at interface with nodes on cylinder end
+        count = 0
+        hemisphereNodesToDelete = []
+        for n2 in range(elementsCountAcross[0] + 1):
+            for n1 in range(elementsCountAcross[1] + 1):
+                if hemisphere.px[elementsCountAcross[2] // 2][n2][n1]:
+                    hemisphereNodesToDelete.append(hemisphere.nodeId[elementsCountAcross[2] // 2][n2][n1])
+                    hemisphere.nodeId[elementsCountAcross[2] // 2][n2][n1] = nodesIdCylinderProximalEnd[count]
+                    count += 1
+
+
+
+        #         # rangeOfRequiredElements = [[0, elementsCountAcross[0]], [0, elementsCountAcross[1]], [0, 1]]
+        #         # nodeIdentifier = max(1, getMaximumNodeIdentifier(nodes) + 1)
+        #         # nodeIdentifier = hemisphere.generateNodes(fm, coordinates, nodeIdentifier,
+        #         #                                           rangeOfRequiredElements)
+        # #**********************************************************************************************************************
+        # generate hemisphere elements.
+        rangeOfRequiredElements = [[0, elementsCountAcross[0]], [0, elementsCountAcross[1]],
+                                   [0, int(0.5 * elementsCountAcross[2])]]
+        elementIdentifier = max(1, getMaximumElementIdentifier(mesh) + 1)
+        elementIdentifier = hemisphere.generateElements(fm, coordinates, elementIdentifier,
+                                                        rangeOfRequiredElements)
+
+        # ***********************************************************************************************************************
+        # top half of the hemisphere
```

Comment:
The code for making the second hemisphere follows the same structure as the ones used in the first hemisphere, with some difference in parameters.  Try combining them into one set of code with a loop instead of having two sets of similar code. It will reduce the length of the code and make your code look more elegant.

Reply:
Will try at the next iteration?